pkg/controller/core/clusterqueue_controller.go

```diff
@@ -497,3 +552,77 @@ func (r *ClusterQueueReconciler) updateCqStatusIfChanged(
 	}
 	return nil
 }
+
+// Taking snapshot of cluster queue is enabled when maxcount non-zero
+func (r *ClusterQueueReconciler) isVisibilityEnabled() bool {
+	return r.queueVisibilityClusterQueuesMaxCount > 0
+}
+
+func (r *ClusterQueueReconciler) getWorkloadsStatus(cq *kueue.ClusterQueue) *kueue.ClusterQueuePendingWorkloadsStatus {
+	if !r.isVisibilityEnabled() {
+		return nil
+	}
+	if cq.Status.PendingWorkloadsStatus == nil {
+		return &kueue.ClusterQueuePendingWorkloadsStatus{
+			Head:           r.qManager.GetSnapshot(cq.Name),
+			LastChangeTime: metav1.Time{Time: time.Now()},
+		}
+	}
+	if time.Since(cq.Status.PendingWorkloadsStatus.LastChangeTime.Time) >= r.queueVisibilityUpdateInterval {
+		pendingWorkloads := r.qManager.GetSnapshot(cq.Name)
+		if !equality.Semantic.DeepEqual(cq.Status.PendingWorkloadsStatus.Head, pendingWorkloads) {
+			return &kueue.ClusterQueuePendingWorkloadsStatus{
+				Head:           pendingWorkloads,
+				LastChangeTime: metav1.Time{Time: time.Now()},
+			}
+		}
+	}
+	return cq.Status.PendingWorkloadsStatus
+}
+
+func (r *ClusterQueueReconciler) Start(ctx context.Context) error {
```

Comment:
what calls this function? Is it internally called within `SetupWithManager`?

It feels a bit like an overload to make the Reconciler also be the one that calls the reconciler? If possible, I would split it into its own struct.

Reply:
I have traced this a little. The `clusterqueue_controller.Start` is called from here: https://github.com/kubernetes-sigs/controller-runtime/blob/2a553d6f910dfc4b0796142ba1e059639a21c54a/pkg/manager/runnable_group.go#L223C17-L223C22.

So, the `Start` is registered in the runnable `Other` group here, and Started here:  https://github.com/kubernetes-sigs/controller-runtime/blob/2a553d6f910dfc4b0796142ba1e059639a21c54a/pkg/manager/internal.go#L417.

The start we call from within Kueue: https://github.com/kubernetes-sigs/kueue/blob/330505e44fd2d6bbfecd708755ba81f59d14906b/cmd/kueue/main.go#L175.

So calling `SetupWithManager` does not call directly the `Start`, but registers the runnable.

I see, seems possible. There could be something like `ClusterQueueSnapshotManager`.

Reply:
Ticketed: https://github.com/kubernetes-sigs/kueue/issues/1099